File: apps/api/src/modules/streaming/protocol.ts
export enum FrameType {
  SCREEN_DELTA = 0x01,
  SCREEN_FULL = 0x02,
  CAMERA = 0x03,
  AUDIO = 0x04,
  HEARTBEAT = 0x05,
  ACK = 0x06,
  CONTROL = 0x07,
}

const FRAME_VERSION = 0x01;
const HEADER_SIZE = 8;

export function parseFrame(buf: Buffer): { type: FrameType; payload: Buffer } {
  if (buf.length < HEADER_SIZE) {
    throw new Error(`Frame too short: ${buf.length} bytes`);
  }
  const type = buf[1] as FrameType;
  const payloadLen = buf.readUInt32BE(4);
  if (buf.length < HEADER_SIZE + payloadLen) {
    throw new Error(`Incomplete frame: expected ${HEADER_SIZE + payloadLen}, got ${buf.length}`);
  }
  return { type, payload: buf.slice(HEADER_SIZE, HEADER_SIZE + payloadLen) };
}

export function buildFrame(type: FrameType, payload: Buffer): Buffer {
  const header = Buffer.alloc(HEADER_SIZE);
  header[0] = FRAME_VERSION;
  header[1] = type;
  header.writeUInt32BE(payload.length, 4);
  return Buffer.concat([header, payload]);
}

export function buildControlFrame(data: Record<string, unknown>): Buffer {
  return buildFrame(FrameType.CONTROL, Buffer.from(JSON.stringify(data)));
}

export function buildHeartbeatFrame(): Buffer {
  return buildFrame(FrameType.HEARTBEAT, Buffer.alloc(0));
}

File: apps/agent/internal/stream/protocol.go
package stream

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
)

const (
	FrameVersion         = byte(0x01)
	FrameTypeScreenDelta = byte(0x01)
	FrameTypeScreenFull  = byte(0x02)
	FrameTypeCamera      = byte(0x03)
	FrameTypeAudio       = byte(0x04)
	FrameTypeHeartbeat   = byte(0x05)
	FrameTypeACK         = byte(0x06)
	FrameTypeControl     = byte(0x07)
	headerSize           = 8
)

// Frame represents a parsed binary frame.
type Frame struct {
	Type    byte
	Payload []byte
}

// BuildFrame creates an 8-byte header + payload binary frame.
// Header layout: [version(1)][type(1)][reserved(2)][payloadLen uint32 BE(4)]
func BuildFrame(frameType byte, payload []byte) []byte {
	header := make([]byte, headerSize)
	header[0] = FrameVersion
	header[1] = frameType
	binary.BigEndian.PutUint32(header[4:], uint32(len(payload)))
	result := make([]byte, headerSize+len(payload))
	copy(result, header)
	copy(result[headerSize:], payload)
	return result
}

// ParseFrame parses a binary frame from raw bytes.
func ParseFrame(data []byte) (*Frame, error) {
	if len(data) < headerSize {
		return nil, fmt.Errorf("frame too short: %d bytes", len(data))
	}
	payloadLen := binary.BigEndian.Uint32(data[4:8])
	if uint32(len(data)) < uint32(headerSize)+payloadLen {
		return nil, fmt.Errorf("incomplete frame: expected %d bytes, got %d", headerSize+payloadLen, len(data))
	}
	return &Frame{
		Type:    data[1],
		Payload: data[headerSize : headerSize+payloadLen],
	}, nil
}

// BuildControlFrame creates a CONTROL frame with JSON payload.
func BuildControlFrame(payload map[string]interface{}) []byte {
	b, _ := json.Marshal(payload)
	return BuildFrame(FrameTypeControl, b)
}

// BuildHeartbeatFrame creates a HEARTBEAT frame with empty payload.
func BuildHeartbeatFrame() []byte {
	return BuildFrame(FrameTypeHeartbeat, []byte{})
}
